File: Programmers/신규아이디추천.py
def solution(new_id):
    new_id = first_step(new_id)
    new_id = second_step(new_id)
    new_id = third_step(new_id)
    new_id = fourth_step(new_id)
    new_id = fifth_step(new_id)
    new_id = sixth_step(new_id)
    new_id = seventh_step(new_id)
    print(new_id)
    answer = new_id
    return answer

def first_step(new_id):
    return new_id.lower()

def second_step(new_id):
    example = ['-', '_', '.']
    for i in new_id:
        if i.isalpha() == False and i.isdigit() == False and i not in example:
            new_id = new_id.replace(i,'')
    return new_id
        
def third_step(new_id):
    stack = []
    for i in new_id:
        if len(stack) == 0:
            stack.append(i)
        else:
            if i == stack[-1] and i == '.' and stack[-1] == '.':
                stack.pop()
                stack.append(i)
            else:
                stack.append(i)
    new_id = ''.join(stack)
    return new_id

def fourth_step(new_id):
    new_id = new_id.strip('.')
    return new_id
        
def fifth_step(new_id):
    if len(new_id) == 0:
        new_id = 'a'
    return new_id

def sixth_step(new_id):
    if len(new_id) >= 16:
        new_id = new_id[0:15]
        if new_id[-1] == '.':
            new_id = new_id.rstrip('.')
    return new_id

def seventh_step(new_id):
    if len(new_id) <= 2:
        count = 3 - len(new_id)
        for i in range(count):
            new_id += new_id[-1]
    return new_id

File: Programmers/신규아이디추천.js
function solution(new_id) {
  const pattern1 = /[0-9]/;
  const pattern2 = /[a-z]/;
  const pattern3 = /[-_.]/;

  //1
  new_id = new_id.toLowerCase();

  //2
  for (let word of new_id) {
    if (!pattern1.test(word) && !pattern2.test(word) && !pattern3.test(word))
      new_id = new_id.replace(word, "");
  }

  //3
  let st_new_id = new Array();
  for (let word of new_id) {
    if (!(word === "." && word === st_new_id.at(-1))) st_new_id.push(word);
  }
  new_id = st_new_id.join("");

  //4
  if (new_id[0] === ".") new_id = new_id.substring(1);
  if (new_id.at(-1) === ".") new_id = new_id.substring(0, new_id.length - 1);

  //5
  if (new_id === "") new_id = "a";

  //6
  if (new_id.length >= 16) {
    new_id = new_id.substring(0, 15);
    if (new_id.at(-1) === ".") new_id = new_id.substring(0, new_id.length - 1);
  }

  //7
  if (new_id.length <= 2) {
    const len = new_id.length;
    for (let i = 0; i < 3 - len; i++) {
      new_id += new_id.at(-1);
      console.log(new_id);
    }
  }
  let answer = new_id;
  return answer;
}
